clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2566,9 +2569,7 @@ class SyclKernelArgsSizeChecker : public SyclKernelFieldHandler {
   bool handleSpecialType(QualType FieldTy) {
     const CXXRecordDecl *RecordDecl = FieldTy->getAsCXXRecordDecl();
     assert(RecordDecl && "The type must be a RecordDecl");
-    llvm::StringLiteral MethodName = (IsSIMD && isSyclAccessorType(FieldTy))
-                                         ? InitESIMDMethodName
-                                         : InitMethodName;
+    llvm::StringLiteral MethodName = InitMethodName;
```

Comment:
> Doesn't this change break ESIMD?

By break, do you mean the ESIMD `InitESIMDMethod` params size getting checked?

Reply:
> Yes, they could be different.

After debugging, it appears that the ESIMD `InitESIMDMethod` params size code was either broken or the concept of kernel arguments size exceeding a supported maximum on certain devices does not apply to ESIMD kernels.

I say already broken because, with the current CFE code, for an example like below, we end up getting an assertion error.

```
void use() {
  struct S {
    int A;
    int B;
    Accessor AAcc;
    Accessor BAcc;
    int Array[1991];
  } Args;
  auto L = [=]()  [[intel::sycl_explicit_simd]]   { (void)Args; };
  q.submit([&](handler &h) {
        h.single_task<class Foo>(L);
  });
}
```


It is also noteworthy that there were no tests written to check ESIMD kernel arguments size exceeding a supported maximum on certain devices (ex: spir64 or spir32). This is probably why we never detected this invalid code.

Given the above reasons, I felt we could remove the ESIMD methodname generation code from the `SyclKernelArgsSizeChecker` Visitor.

Please let me know your thoughts.

Reply:
> Given the above reasons, I felt we could remove the ESIMD methodname generation code from the SyclKernelArgsSizeChecker Visitor. We should probably add code to check the ESIMD kernel arguments size limit too (if there is one), but we can do it as a separate patch.

Could you please double check this with ESIMD team? And if the diagnostic is required, please create an internal tracker.

Reply:
It appears the ESIMD specifics are totally similar to SYCL, except that the sub-group-size =1.
I will open a tracker to fix the gap.

Reply:
>>I say already broken because, with the current CFE code, for an example like below, we end up getting an assertion error.

What was the assertion you hit? What happens now that you've changed the code to use __init and not __init_esimd or whatever the ESIMD init function is called.

Reply:
> > > I say already broken because, with the current CFE code, for an example like below, we end up getting an assertion error.
> 
> What was the assertion you hit? What happens now that you've changed the code to use __init and not __init_esimd or whatever the ESIMD init function is called.

You hit this assertion [here](https://github.com/intel/llvm/blob/sycl/clang/lib/Sema/SemaSYCL.cpp#L2573) when you try checking the kernel arguments size for ESIMD kernels.

This is because [getMethodByName](https://github.com/intel/llvm/blob/sycl/clang/lib/Sema/SemaSYCL.cpp#L959) is null, as there were no matching methods by name - `__init_esimd`

I haven't added any new code to change this behavior. I have simply removed the ESIMD code leading to this assertion.
Does this make sense?

Reply:
Were we hitting the assertion on a valid esimd program?

Reply:
That....makes no sense. If the assert is being hit for `ArgChecker`, it means it should be hit for other visitors. If that is the case, I'm assuming the accessor definition in mock header is incorrect

Reply:
> Were we hitting the assertion on a valid esimd program?



> That....makes no sense.

Ok, let me try again.

With the current CFE code, I hit an assertion with a valid ESIMD program, when the ESIMD kernel had `accessors `as it's arguments.

This assertion was because, the `__init` variant used by the device compiler for ESIMD kernels, `__init_esimd`, was never found in the `accessor` class definition.

Also, the assertion in `SyclKernelArgsSizeChecker` occurs much earlier in the pipeline, before `SyclKernelDeclCreator` and 
`SyclKernelBodyCreator`, where the same assertion should also occur.

Because of this, I never got around reproducing the assertion in `SyclKernelDeclCreator` and 
`SyclKernelBodyCreator`

Now, I encountered the said assertions while running an existing test program, which I modified to an ESIMD program. The test was using the SYCL mock headers.

Admittedly, I did not consider that the implementation of accessor class in the mock headers could be missing the inclusion of __init_esimd method for accessors. (reason #862863325654 to not trust mock header implementation - note to self)

Considering there were no tests written to check ESIMD kernels argument size and me hitting an assertion using a valid ESIMD kernel code, made me believe the code handling argument checking could be dubious.

It indeed turns out that the mock header file in clang/test/SemaSYCL, is missing the __init_esimd method in accessor class, and adding it makes the assertion go away.

So, thank you, @elizabethandrews for bringing this up!

Since we are moving ESIMD logic from `CheckSYCLKernelCall` to `ConstructOpenCLKernel`, we could handle this case by moving the ESIMD argument size checking to `SyclKernelDeclCreator`
WDYT?

Reply:
@Fznamznon Could you please weigh in your thoughts here as well?

Reply:
I don't have a strong preference, but keeping this as a separate visitor called from `ConstructOpenCLKernel` seems less invasive.

Reply:
I don't know... We're adding more diagnostics into DeclCreator which makes me uncomfortable. Does a separate checker for ESIMD diagnostics make sense in ConstructOpenCLKernel? It 'feels' better than adding more diagnostics to DeclCreator to me but maybe the other reviewers have a different opinion